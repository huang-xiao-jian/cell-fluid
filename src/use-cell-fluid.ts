/**
 * @description - format payload while typing
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// package
import { useEffect, useState, useCallback, ChangeEvent } from 'react';
import { Subject, Subscription } from 'rxjs';
import { map } from 'rxjs/operators';

export interface CellFluidOptions {
  accept: RegExp;
  format: (payload: string) => string;
}

/**
 * @description - cleanup additional format symbols
 *
 * @param {string} value
 * @param {RegExp} accept
 */
export function clean(value: string, accept: RegExp = /\d/g): string {
  return (value.match(accept) || []).join('');
}

/**
 * @description - assume actual payload is identical after format
 *
 * @param {string} before - user input without additional symbols
 * @param {string} after - formatted payload
 */
export function diff(before: string, after: string) {
  let position = 0;
  const length = before.length;

  for (let i = 0; i < length; i++) {
    position = after.indexOf(before[i], position) + 1;
  }

  return position;
}

/**
 * @description - handle payload format and cursor shift
 *
 * @param {CellFluidOptions} options
 */
export function useCellFluid(
  options: CellFluidOptions
): [string, (event: ChangeEvent<HTMLInputElement>) => void] {
  const [value, setValue] = useState('');
  const [position, setPosition] = useState(0);
  const [target, setTarget] = useState<HTMLInputElement | null>(null);
  const [events$] = useState(new Subject<ChangeEvent<HTMLInputElement>>());
  const handleValueChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => events$.next(event),
    []
  );

  useEffect(() => {
    const subscription: Subscription = events$
      .pipe(
        map((event) => {
          const payload = event.currentTarget.value;
          const payloadBeforeCursor = clean(
            payload.slice(0, event.currentTarget.selectionStart as number)
          );
          const formatted = options.format(payload);

          return {
            formatted,
            position: diff(payloadBeforeCursor, formatted),
            target: event.currentTarget as HTMLInputElement,
          };
        })
      )
      .subscribe((state) => {
        // reflect user value
        setValue(state.formatted);

        // prepare for correct cursor position
        setTarget(state.target);
        setPosition(state.position);
      });

    return () => {
      // complete callback
      events$.complete();
      // cancel event stream
      subscription.unsubscribe();
    };
  }, []);

  useEffect(() => {
    if (target) {
      target.setSelectionRange(position, position);
    }
  }, [position, target]);

  return [value, handleValueChange];
}
