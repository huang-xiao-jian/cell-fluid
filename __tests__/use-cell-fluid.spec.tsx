/**
 * @description - test suits
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// package
import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { clean, diff, useCellFluid } from '../src/use-cell-fluid';

function Fixture() {
  const [value, handleValueChange] = useCellFluid({
    accept: /\d/g,
    format: (raw: string) =>
      raw.replace(/\s/g, '').replace(/(\d{4})(?=\d)/g, '$1 '),
  });

  return (
    <input data-testid="fixture" value={value} onChange={handleValueChange} />
  );
}

describe('useCellFluid hook', () => {
  it('should clean additional symbols', () => {
    expect(clean('1a34 ab5')).toEqual('1345');
    expect(clean('1a3x ab5', /[x\d]/g)).toEqual('13x5');
  });

  it('should calculate right cursor position', () => {
    expect(diff('1234', '1234')).toEqual(4);
    expect(diff('123456', '1234 56')).toEqual(7);
  });

  it('should render with hook', async () => {
    const { container, findByTestId } = render(<Fixture />);
    const input = (await findByTestId('fixture')) as HTMLInputElement;

    expect(container.innerHTML).toMatchSnapshot();

    fireEvent.change(input, {
      target: {
        value: '12345',
        selectionStart: 5,
      },
    });

    expect(container.innerHTML).toMatchSnapshot();
    expect(input.selectionStart).toEqual(6);
  });
});
