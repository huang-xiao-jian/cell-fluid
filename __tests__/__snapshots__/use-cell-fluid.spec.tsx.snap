// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCellFluid hook should render with hook 1`] = `"<input data-testid=\\"fixture\\" value=\\"\\">"`;

exports[`useCellFluid hook should render with hook 2`] = `"<input data-testid=\\"fixture\\" value=\\"1234 5\\">"`;
